### ISSUE TITLE
Desync for Australian/New Zealand users on US headless sessions with high player count

### ISSUE THREAD

**Issue by Crusher50** [Neos-Metaverse/NeosPublic#3039]
### Describe the bug?

For primarily Australian and New Zealand users that are on a US headless servers with a higher player count (around 8 to 10 or more) Desync begins to occur.

Note that desync does begin to reduce as the amount of users in the session reduce (is playable at around 4 to 5 or lower).

Desync begins to affect primarily:
Inspector panels taking 100 to 200 seconds to show any slots.
 
objects other users have grabbed only catching up after around 100 seconds (the same time it takes for inspectors to catch 
up).

other users voice modes. for example if someone, that is not the desynced user, changes their voice mode to mute they will not be muted for the desynced user until the desync has catches up which they will be muted mid sentence.  
(note that their is no desync between peer to peer, like talking to another user is not affected by desync as well as their avatar's movements)

objects other users have spawned only appearing for the desynced user once desync catches up.

any changes other users have made, like changing a material, mesh etc.

(this issue has only started from today) when a user changes their avatar, they appear frozen until the desync catches up. This also stops all peer to peer interaction that hasn't normally been affected by desync like talking and avatar movement.

### Relevant issues

no apparent relevant issues that haven't already been fixed.

### To Reproduce

Reproducing this issue may be hard to replicate for users other than Australian and New Zealand users but is a consistent issue in creator jams where the player count is at around 10 or more and the headless is US.

the main things needed to reproduce the bug are to be connect to a headless that is generally a very far distance from where you a playing (a VPN may help with this) and to have the session have 10 or more users.

### Expected behavior

to have no desync in headless that are in a different region to users that join, and for there to be no desync with a high player count session.

### Log Files

[DESKTOP-F65B6FQ - 2021.9.16.105 - 2021-09-19 10_57_51.log](https://github.com/Neos-Metaverse/NeosPublic/files/7192024/DESKTOP-F65B6FQ.-.2021.9.16.105.-.2021-09-19.10_57_51.log)

the issue occurs at 1:28:57 PM.179

### How often does it happen?

Always

### Does the bug persist after restarting Neos?

Yes

### Neos Version Number

2021.9.16.105

### What Platforms does this occur on?

Windows, Linux

### Did this work before?

I Don't Know

### Reporters

Neos: Crusher
Discord: Crusher#6146

**Comment by shadowpanther** [comment]
As far as I know, Australia has a very low-bandwidth uplink to the rest of the world, so even if your local connection has high bandwidth, connection to any host outside of Australia would be rate-limited. Desync happens because that rate-limit is saturated with streams from all users coming through the host user. Datamodel changes have lower priority and thus changes get delayed.

I'm not sure what the solution to this would be apart from adding more uplink cables from Australia to the rest of the world to add bandwidth. Maybe [mesh networking](https://github.com/Neos-Metaverse/NeosPublic/projects/13#card-61574582) would help a bit as you would have multiple connections to different hosts instead of one.

**Comment by Crusher50** [comment]
That does make sense, although the only thing that I would believe has a higher priority is the voice mode over other users but this doesn't seem to be the case.

**Comment by ohzee00** [comment]
Sadly I'm unsure how much can be done here(besides perhaps mesh networking linked above) however the only possible suggestion is asking users to change their settings to Steam networking sockets in Neos. 
Note though, this only works if the host has steam being it runs through their service, headlesses might need to be configured with that in mind.

Mind you this isn't a direct fix, I just remember helping some users before that were desyncing badly due to latency and that Steam networking sockets helped them somewhat, at least allowed them to play in a semi-usable state.

**Comment by Frooxius** [comment]
Unfortunately this is a limitation of the current transport protocol that we are using, it unfortunately degrades with certain connections (typically high latency, but can also be just a result of quirks of given connection).

You can try switching to "Prefer Steam Networking Sockets" in the settings, which can behave a lot better in these scenarios, but unfortunately currently doesn't have bandwidth estimation, so it can end up dropping the connection as well.

We are currently waiting on Valve to implement this feature so we can switch to it as the main protocol (or specifically the open version Game Networking Sockets).

**Comment by Enverex** [comment]
Would this also be why people on very low speed connections (e.g. ~3Mb/s down, 0.5Mb up ) are unable to ever reach sync, even in a world with one other person? I have a friend in France and regardless of whether he is host or client, even with a single other person present with almost nothing else happening, he will not be able to maintain sync.

**Comment by shadowpanther** [comment]
> doesn't have bandwidth estimation

Speaking of, if bandwidth estimation would be implemented (by SNS or GNS), could it be used by Neos to downgrade audio streams bitrate and maybe pose stream frequency to allow for eventual sync for clients with very limited bandwidth?

**Comment by iamgreaser** [comment]
GreaseMonkey here.

The server we were getting desynced on is the main Creator Jam hub server which is hosted in Ukraine. I get a ~305 msec ping to it.

For reference, about 8 years ago I was getting pings of about 200 msec to US West, 260 msec to US East, and 350 msec to Europe. Nowadays I get about 180 msec to US West and I haven't remeasured the rest.

I think I'm hitting the LNL Relay connection in several cases, even though I know that my NAT handles UDP holepunching just fine.

Connection speed is... pretty decent? I suspect it does boil down to latency.

**Comment by Lexevolution** [comment]
Whenever I had any major desync on some sessions, and switched to SNS, I sometimes got this strange side effect where everything seemed like it was synced, but in realty, all of my actions, voice and my perspective of the world was 30+ seconds behind. And to everyone else, I was lagging behind their conversations/actions by 30+ seconds. It looked very different to the regular desyncing issue which doesn't usually affect voices.

**Comment by Frooxius** [comment]
@Enverex Don't know, we'd need to gather data on that. Does Steam Networking Sockets make a difference? Could be lots of things with the connection.

@shadowpanther That's unlikely, typically that detail is hidden in the protocol itself and not accessible. Generally those don't pose much of an issue anyways, since those are using streams and can be lost. It's the reliable changes that start queuing up. Usually it's not even the bandwidth itself, but rather packet rate.

@iamgreaser Sometimes it's the combination of connections that just don't work with UDP holepunching. We've had cases where you could have connection A, B and C. A and B would work fine, B and C would work fine with each other, but A and C will always go through the relay.

@Lexevolution Steam Networking Socket handles things differently, which is why you get that effect. Essentially you hit the fixed bandwidth limit, so everything starts trickling through and delaying like that. It's why we need the bandwidth estimation to be implemented in the protocol before we can switch to it as primary one.

**Comment by sveken** [comment]
Is there any roadmap or really rough date when we could see that happening? Unfortunately the desync issue for Australians which i only experience on Neos pretty much locks us out of things like the MetaMovie which was a ton of fun when it worked and other big events,

**Comment by Crusher50** [comment]
quick update on this, Steam Networking helped significantly (only 1 to 3 seconds of desync but it's more latency at this point) though it does need to be enabled on the headless first. It has also only disconnected me once when around 10 fairly heavy assets were deleted.

**Comment by Frooxius** [comment]
@sveken We asked on the bandwidth estimation here: https://github.com/ValveSoftware/GameNetworkingSockets/issues/108 But currently there's no ETA on their end so we just have to wait and see before the switch. I saw some movement for bandwidth estimation a few months ago.

There are a few things that I'm looking into on our end though that might help improve the network performance, mainly with combining smaller messages into a bigger one to reduce the overall packet-rate.

**Comment by sveken** [comment]
Just did the MetaMovie again, 
The starting was much better only starting what felt like 5 seconds of dysync compared to 5 minutes before the update.
However further along in the story the deysnc did get worse as more things happened, i was told my high score was 50K queued packets. 
Bandwidth did not go over 1.3mb/s on the download for Neos. 
World had 14 people in it. 

Definitely an improvement however.

**Comment by rabbuttz** [comment]
I don't know this issue is related to this, but I have recently been experiencing out-of-sync issues in certain sessions. This has been happening more frequently since the recent update regarding the network came in. Specifically, item locations, Logix processing, etc. are out of sync. Strangely enough, the voice and user locations seem to be perfectly in sync, and when I check the QueuedPackets in Neos, it shows 0. Sometimes I can't see the user even though they are supposed to be there. When I go to the dash menu and look at the session details, it looks like the user is not there. This problem did not seem to be related to whether or not I was using SteamNetworkingSockets. My internet speed is 368.0Mbps↓/29.5Mbps↑. The logs are attached. The problem occurs when I'm in a session called "SLOT開発室", around 3:00.
https://www.dropbox.com/s/69oi5kvwi3tvj5f/DESKTOP-CB6HJR0%20-%202021.10.30.625%20-%202021-11-01%2020_12_06.log?dl=0
